### PR TITLE
support flashinfer_fp4 moe for 5090 gpu

### PR DESCRIPTION
**Pull request by XiaobingSuper**
## Purpose
This PR is to enable lashinfer_fp4 to work for 5090 gpu device.

## Test Plan

```python

import numpy
import torch

from vllm import LLM, SamplingParams


prompts = [
    "The Swiss Alps are", 
    "Brad Marchand is",
    "The Boston Bruins are"
]


sampling_params = SamplingParams(temperature=0.90, top_p=0.95, max_tokens=40, min_tokens=10)
llm = LLM("nv-community/Qwen3-30B-A3B-FP4", tensor_parallel_size=1, enforce_eager=True)

responses = llm.generate(prompts, sampling_params)
for output in responses:
    prompt = output.prompt
    generated_text = output.outputs[0].text
    print(f"Prompt: {prompt!r}, Generated text: {generated_text!r}")

```

```bash
VLLM_USE_FLASHINFER_MOE_FP4=1 VLLM_USE_V1=1 python qwen_fp4_test.py
```

## Test Result

```bash 
Prompt: 'The Swiss Alps are', Generated text: ' a mountain range in which country?\n\nThe Swiss Alps are a mountain range located in Switzerland. They are part of the larger Alps mountain system, which stretches across several European countries. The Swiss Alps are particularly'
Prompt: 'Brad Marchand is', Generated text: ' one of the best goal scorers in the NHL. But he’s also one of the best at drawing penalties. For the second straight season, he’s leading the league in penalty minutes. But it'
Prompt: 'The Boston Bruins are', Generated text: ' a professional ice hockey team based in Boston, Massachusetts. They are members of the Atlantic Division of the Eastern Conference in the National Hockey League (NHL). The team was founded in 192'
```

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

### CODE REVIEW

**Review by yewentao256 — APPROVED** [review]
LGTM, thanks for the work!

File: vllm/model_executor/layers/quantization/utils/flashinfer_fp4_moe.py [hunk 1]
```diff
@@ -32,7 +32,7 @@ def is_flashinfer_fp4_cutlass_moe_available() -> bool:
         envs.VLLM_USE_FLASHINFER_MOE_FP4
         and has_flashinfer_cutlass_fused_moe()
         and current_platform.is_cuda()
-        and current_platform.is_device_capability(100)
+        and any(current_platform.is_device_capability(cap) for cap in [100, 120])
```
Comment (jasl) [inline vllm/model_executor/layers/quantization/utils/flashinfer_fp4_moe.py]: Can you add 110 (Thor) and 121 (Spark)?

Comment (XiaobingSuper) [reply vllm/model_executor/layers/quantization/utils/flashinfer_fp4_moe.py]: added.

**Comment by PlayHard4Juice** [comment]
>the problem is that I got exactly the same speed in FP4 as FP8 which means native NVFP4 is not used. It must be theoretically 2x faster - practically at least faster. I have RTX 6000 PRO so the FP8 full model fits

Maybe this is related?
https://github.com/NVIDIA/TensorRT-LLM/issues/4412

they say in TensorRT-LLM,
>Compared to FP8, FP4 shows faster performance during the prefill stage, but FP4 is actually slightly slower during the decode stage.

So maybe we should test the prefill speed as well

**Review by jasl — APPROVED** [review]
Thank you! cc @johnnynunez

**Comment by PlayHard4Juice** [comment]
I tested it again and this time I looked at the response content. It responses nonsense sometimes, and sometimes it loops forever. It's prompt and random seed related. For some prompt it's ok, but sometimes it's not.

Model: NVFP4/Qwen3-30B-A3B-Instruct-2507-FP4

By the way, TensorRT-LLM's implementation also has this behavior. Something must be wrong. Maybe it should be another issue.

File: vllm/model_executor/layers/quantization/utils/flashinfer_fp4_moe.py [hunk 2]
```diff
@@ -32,7 +32,9 @@ def is_flashinfer_fp4_cutlass_moe_available() -> bool:
         envs.VLLM_USE_FLASHINFER_MOE_FP4
         and has_flashinfer_cutlass_fused_moe()
         and current_platform.is_cuda()
-        and current_platform.is_device_capability(100)
+        and any(
+            current_platform.is_device_capability(cap) for cap in [100, 110, 120, 121]
+        )
```
Comment (mgoin) [inline vllm/model_executor/layers/quantization/utils/flashinfer_fp4_moe.py]: switching to has is sufficient here for greater than or equal
```suggestion
        and current_platform.has_device_capability(100)
```

**Review by mgoin — APPROVED** [review]
